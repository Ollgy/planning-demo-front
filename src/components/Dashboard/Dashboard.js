import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import SectionTitle from '../SectionTitle';
import TaskList from '../TaskList';
import NewTaskForm from '../NewTaskForm';
import Cap from '../Cap';
import Preloader from '../Preloader';
import { CSSTransition } from 'react-transition-group';
import {
  setDialog,
  setCurItem,
  deleteTask,
  updateTask,
  getTasks
} from '../../modules/Dashboard';
import { D, P, M } from '../../const';
import styles from './Dashboard.module.css';

class Dashboard extends PureComponent {
  componentWillUnmount() {
    const { newTaskWindow } = this.props.dialog;
    const { setCurItem } = this.props;

    setCurItem(null);
    if (newTaskWindow) {
      this.closeTaskWindow();
    }
  }

  async componentDidMount() {
    const { userId, getTasks } = this.props;

    await getTasks(userId);
  }

  openTaskWindow = (mode, data) => {
    const { setCurItem, setDialog } = this.props;

    setCurItem(data);
    setDialog({ dialog: D.newTaskWindow, value: true });
  };

  closeTaskWindow = () => {
    const { setDialog } = this.props;

    setDialog({ dialog: D.newTaskWindow, value: false });
  };

  deleteTask = async taskId => {
    const { deleteTask } = this.props;

    await deleteTask(taskId);
  };

  updateTask = async (taskId, taskData) => {
    const { updateTask } = this.props;

    await updateTask(taskId, taskData);
  };

  render() {
    const {
      userPermission,
      userName,
      dialog,
      tasks,
      curTaskItem,
      isLoading
    } = this.props;
    const { newTaskWindow } = dialog;

    return !isLoading ? (
      <section className={styles.sectionContainer}>
        <SectionTitle
          children={
            userPermission === P.manager ? 'Список задач' : 'Мои задачи'
          }
          size={userPermission === P.manager ? 's' : 'm'}
        />
        {userPermission === P.manager ? (
          <span className={styles.subtitle}>{`Сотрудник: ${userName}`}</span>
        ) : null}
        {tasks.length ? (
          <TaskList
            list={tasks}
            userPermission={userPermission}
            activities={{
              updateTask: this.updateTask,
              deleteTask: this.deleteTask,
              openTaskWindow: this.openTaskWindow.bind(null, M.update)
            }}
          />
        ) : (
          <Cap text="Текущих задач не найдено" />
        )}
        <CSSTransition
          in={newTaskWindow}
          timeout={300}
          mountOnEnter={true}
          unmountOnExit={true}
          classNames={{
            appear: styles.appear,
            enter: styles.enter,
            enterActive: styles.enterActive,
            exit: styles.exit,
            exitActive: styles.exitActive
          }}
        >
          <NewTaskForm
            mode={M.update}
            taskData={curTaskItem}
            updateTask={this.updateTask}
            onClose={this.closeTaskWindow}
          />
        </CSSTransition>
      </section>
    ) : (
      <div className={styles.preloaderBlock}>
        <Preloader />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  dialog: state.dashboard.datalist.dialog,
  curTaskItem: state.dashboard.datalist.curTaskItem,
  tasks: state.dashboard.datalist.tasks,
  isLoading: state.dashboard.getTasks.isLoading
});

const mapDispatchToProps = dispatch => ({
  updateTask: (arg1, arg2) => dispatch(updateTask(arg1, arg2)),
  deleteTask: arg => dispatch(deleteTask(arg)),
  getTasks: arg => dispatch(getTasks(arg)),
  setCurItem: arg => dispatch(setCurItem(arg)),
  setDialog: arg => dispatch(setDialog(arg))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Dashboard);
