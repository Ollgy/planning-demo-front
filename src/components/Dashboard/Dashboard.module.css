.sectionContainer {
  padding-top: 20px;
  overflow-y: auto;
}

.subtitle {
  display: inline-block;
  margin-bottom: 20px;
}

.preloaderBlock {
  height: 250px;
}
