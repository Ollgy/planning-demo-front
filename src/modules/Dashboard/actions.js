import { createAction } from 'redux-actions';
import api from '../../api';

export const setCurItem = createAction('SET_CUR_TASK_ITEM');
export const setDialog = createAction('SET_DIALOG_TASK');

export const updateTaskRequest = createAction('UPDATE_TASK_REQUEST');
export const updateTaskSuccess = createAction('UPDATE_TASK_SUCCESS');
export const updateTaskFailure = createAction('UPDATE_TASK_FAILURE');

export const updateTask = (taskId, taskData) => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(updateTaskRequest());

      api
        .updateTask(taskId, taskData)
        .then(tasks => {
          dispatch(updateTaskSuccess(tasks));
          resolve();
        })
        .catch(error => {
          dispatch(updateTaskFailure(error));
          reject();
        });
    });
};

export const deleteTaskRequest = createAction('DELETE_TASK_REQUEST');
export const deleteTaskSuccess = createAction('DELETE_TASK_SUCCESS');
export const deleteTaskFailure = createAction('DELETE_TASK_FAILURE');

export const deleteTask = taskId => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(deleteTaskRequest());

      api
        .deleteTask(taskId)
        .then(tasks => {
          dispatch(deleteTaskSuccess(tasks));
          resolve();
        })
        .catch(error => {
          dispatch(deleteTaskFailure(error));
          reject();
        });
    });
};

export const getTasksRequest = createAction('GET_TASKS_REQUEST');
export const getTasksSuccess = createAction('GET_TASKS_SUCCESS');
export const getTasksFailure = createAction('GET_TASKS_FAILURE');

export const getTasks = userId => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(getTasksRequest());

      api
        .getTasks(userId)
        .then(tasks => {
          dispatch(getTasksSuccess(tasks));
          resolve();
        })
        .catch(error => {
          dispatch(getTasksFailure(error));
          reject();
        });
    });
};
