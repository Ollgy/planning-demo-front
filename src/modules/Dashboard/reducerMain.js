import { combineReducers } from 'redux';
import delTask from './reducers/delTask';
import updTask from './reducers/updTask';
import getTasks from './reducers/getTasks';
import datalist from './reducers/datalist';

export default combineReducers({
  datalist,
  getTasks,
  delTask,
  updTask
});
