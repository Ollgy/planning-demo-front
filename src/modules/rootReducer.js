import { combineReducers } from 'redux';
import authFromToken from './Auth';
import login from './Login';
import loginForm from './LoginForm';
import profile from './Profile';
import personalForm from './PersonalForm';
import changePswrdForm from './ChangePswrdForm';
import newTaskForm from './NewTaskForm';
import notes from './Notes';
import dashboard from './Dashboard';
import staff from './Staff';
import cleanBase from './CleanBase';

export default combineReducers({
  authFromToken,
  login,
  loginForm,
  profile,
  personalForm,
  changePswrdForm,
  newTaskForm,
  notes,
  dashboard,
  staff,
  cleanBase
});
