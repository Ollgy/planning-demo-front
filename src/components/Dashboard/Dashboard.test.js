import React from 'react';
import Dashboard from './Dashboard';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { createStore, compose, applyMiddleware } from 'redux';
import rootReducer from '../../modules/rootReducer';
import thunk from 'redux-thunk';

const store = createStore(rootReducer, {}, compose(applyMiddleware(thunk)));

describe('<Dashboard/> UI component', () => {
  const shallowExpect = compose(
    expect,
    toJSON,
    shallow
  );

  it('Renders correctly', () => {
    shallowExpect(<Dashboard store={store} />).toMatchSnapshot();
  });

  const wrapper = shallow(<Dashboard store={store} />);
  const inner = wrapper.children().first();

  it('contains tasks in props', () => {
    expect(inner.props().tasks).toBeTruthy();
  });
});
