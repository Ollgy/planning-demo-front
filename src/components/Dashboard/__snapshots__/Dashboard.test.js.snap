// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard/> UI component Renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <Dashboard
    curTaskItem={Object {}}
    deleteTask={[Function]}
    dialog={false}
    getTasks={[Function]}
    isLoading={false}
    setCurItem={[Function]}
    setDialog={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    tasks={Array []}
    updateTask={[Function]}
  />
</ContextProvider>
`;
